### PR TITLE
[clang-tidy] Add new modernize-use-starts-ends-with check

### PR DESCRIPTION
**Pull request by nicovank**
Make a modernize version of abseil-string-find-startswith using the available C++20 `std::string::starts_with` and `std::string_view::starts_with`. Following up from https://github.com/llvm/llvm-project/pull/72283.

Possible reviewers: @PiotrZSL @njames93

### CODE REVIEW

**Comment by firewave** [comment]
I wonder if this should also detect the `str.compare("marker", 0, 6) == 0` pattern. There is possibly some kind of pattern involving `std::equal()` as well. Could as well be a different check though.

Not sure if it would have a performance impact to use `starts_with()` instead though.

**Comment by PiotrZSL** [comment]
would be to support also endswith in same check, so i'm not so sure about check name.

File: clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp:77]: use diffrent binding instead, simplyone bind to findfun, other to rfindfun

File: clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp:84]: Wont work with macros, you may want to get expansion location

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp:84]: Other workaround could be to use inserts instead of replacement

File: clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.h
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.h:34]: use TK_IgnoreUnlessSpelledInSource, check other checks...

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:189]: "This is both a readability and a performance issue." no need for that in release notes

File: clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp
Comment (nicovank) [reply clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp:84]: I thought that lines 73-75 were checking and ignoring macro expansions?

```
if (ComparisonExpr->getBeginLoc().isMacroID()) {
  return;
}
```

**Comment by firewave** [comment]
> would be to support also endswith in same check

+1

On a side note: I will be looking into the related patterns and their performance soon as I am getting very strange code/performance when they are used outside of a benchmark - especially with Clang.

**Comment by nicovank** [comment]
> Detect the `str.compare("marker", 0, 6) == 0` pattern.

This is actually in my notes for making this check / codemod. No performance impact so not tackling yet, maybe in the future.

> Support `ends_with` in same check.

Similar as above, matching for that length argument is a bit trickier than just looking for 0. Not tackling for now. Any thoughts on open-ended check name instead? `modernize-string-find-affix` (affix = prefix | suffix)?

File: clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.h
Comment (nicovank) [reply clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.h:34]: Was this always recommended? Maybe it should be added to the `add_new_check` template... Along with `isLanguageVersionSupported`.

**Comment by PiotrZSL** [comment]
Maybe modernize-use-starts-ends-with

**Comment by nicovank** [comment]
I like `XXX-use-starts-ends-with`. I would rather move it to `performance` than `readability` if not `modernize`. @PiotrZSL thoughts before I go ahead and do the renaming? Also move it somewhere else?

**Review by EugeneZelenko — CHANGES_REQUESTED** [review]

**Comment by PiotrZSL** [comment]
Performance is fine, go ahead.

**Comment by nicovank** [comment]
Addressed feedback, renamed to `performance-use-starts-ends-with`, relaxed use of `auto`.

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:31]: Check this other matcher that we have, the one from utils directory, that is already used for example in enum-size check

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:38]: this wont work with class inheritence, if someone derive from std::string, instead you could use:
callee(cxxMethodDecl(ofClass(hasAnyName....)))

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:38]: Also in same place you could verify if class got also starts_with and ends_wth method, in such case classes like llvm::StringRef or some boost one could be supported at hock.

File: clang-tools-extra/clang-tidy/abseil/StringFindStartswithCheck.h
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/abseil/StringFindStartswithCheck.h:37]: Use TK_IgnoreUnlessSpelledInSource to exclude things like template instances and  implicit code.

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:83]: if possible avoid using getSourceText this mess out with multiple fixes.
Consider just using multiple CreateReplacement and CreateInsertion

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:84]: #define STR str
if Haystack is from macro:
```
  std::string STR;
  if (0 == STR.find("ala"))
```

then it won't work.

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:93]: actually you could point here to place where call to find/rfind happen

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:97]: may cause problem with comparison ends with macro

File: clang-tools-extra/docs/clang-tidy/checks/abseil/string-find-startswith.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/abseil/string-find-startswith.rst:12]: use :doc:`performance-use-starts-ends-with <../performance/use-starts-ends-with>`

File: clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/modernize/StringFindStartswithCheck.cpp]: only begin location

**Comment by nicovank** [comment]
Major changes:
 -  Now matching on any class that has a `starts_with`, `startsWith`, or `startswith` function. This is done in order (prioritising functions in subclasses, but this could be changed). Experimenting on making this an option, not sure it's worth it without controlling class names as well.
 -  Not using `Lexer::getSourceText` anymore.

File: clang-tools-extra/test/clang-tidy/checkers/Inputs/Headers/stddef.h
Comment (nicovank) [inline clang-tools-extra/test/clang-tidy/checkers/Inputs/Headers/stddef.h:15]: This was just issuing a warning in some tests.

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (nicovank) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:29]: Using `hasAnyName` will match the first declared function. With this setup there is a precedence, I thought this was better. IIRC `llvm::StringRef` has `startswith` and `starts_with` and the latter is preferred.

This could maybe be an option, e.g. `OrderedStartsWithFunctions`. The code gets a bit trickier because this is the opposite of `mapAnyOf`, and AFAIK `anyOf` doesn't take a vector or other kind of collection, only varargs. Suggestions welcome.

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:41]: consider putting this anyOf first, to exclude lot of calls that have more than 1 argument

File: clang-tools-extra/test/clang-tidy/checkers/performance/use-starts-ends-with.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/performance/use-starts-ends-with.cpp:99]: Add few more test withs macros:
- method name under macro
- argument under macro
- 0 under macro
- operator under macro  - #define EQ(x,y) (x) == (y)

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.h
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.h:28]: We may not need this limitation anymore...
Simply llvm::StringRef will also work, and that's C++17.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.h:28]: Limit it to C++ only.

**Review by PiotrZSL — COMMENTED** [review]
Looks good. Minor cleanup and could land.

File: clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:29]: You may verify if those methods are also const.

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/performance/UseStartsEndsWithCheck.cpp:29]: And verify number of argument that they provide.

**Comment by nicovank** [comment]
Updates. I've been running into issues when the begin location of the entire expression is a macro. The prefix removal, which should include the macro text, is not working as I expect it to. If you know of any checks that have similar logic I can take a look and copy. I just disabled it if `ComparisonExpr->getBeginLoc().isMacroID()`. This is what the original check also did so I imagine they had a similar issue.

**Comment by EugeneZelenko** [comment]
I think `modernize` is better fit for this check.

**Comment by nicovank** [comment]
@EugeneZelenko We've had this discussion, see here and the couple comments below: https://github.com/llvm/llvm-project/pull/72385#issuecomment-1812801402. I guess I can rename one more time if there is consensus.

**Comment by EugeneZelenko** [comment]
@firewave, @carlosgalvezp: What do you think?

**Comment by PiotrZSL** [comment]
Problem that I see with "modernize" category is that it overlap with other categories, for example modernize-make-shared is mainly a performance check, modernize-redundant-void-arg is an readability check. We basically lack category for code design (rule of 5, limiting class/file size, complexity), we lack category for security issues (things that are clearly a bug). And we got generic checks in some codding standard categories, that almost every project may need to enable like google-explicit-constructor. If we would have actual check aliases or check tags, so some checks could be target for example with "google" or "cert" tag then it would be different story. But with current speed of delivering changes to clang-tidy, it will never happen.

**Comment by EugeneZelenko** [comment]
Historically, `modernize` was separate binary which was merged with `Clang-tidy`. Its intend was to simplify migration to features added in newer versions of C++. Even if performance/readability are affected, code migration to newer standards is primary advantage.

**Comment by nicovank** [comment]
Ping. @carlosgalvezp could you please weigh in? modernize / performance? Anyone else?

I have no real preference either way. Now that modernize is fully integrated in Clang-Tidy, I don't think historical reasons should lead to modernize overbloat. If this check was only replacing `strncmp` with `starts_with` it would definitely belong in modernize, as-is it sits in-between.

**Comment by carlosgalvezp** [comment]
Apologies for the late reply, need to fix my mail filter so I don't miss the notifications!

IMHO either modernize or readability are suitable categories. Performance can vary/be negligible depending on context I suppose?

**Comment by nicovank** [comment]
Thank you!!
Modernize wins the vote, I've renamed the check.

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-starts-ends-with.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-starts-ends-with.rst:4]: make it shorter to match name

**Review by PiotrZSL — APPROVED** [review]
For a starts_with looks fine, consider now adding support for ends_with, maybe in separate push request once this will be merged.

**Comment by nicovank** [comment]
> Support for `ends_with`.

Planning to look into it and other `starts_with` patterns.

I do not have commit access, please merge for me (`Nicolas van Kempen <nvankempen@meta.com>`).

**Comment by PiotrZSL** [comment]
@nicovank I run check on llvm, reported 3 issues, no false-positives. There is warning reported for FindFun variable because is used only in assert, check if you could do something with it, maybe add if for checking agains null, to silent warning or removing this variable and assert.
Check your github config (email privacy) otherwise this PR will be merged as nicovank@users.noreply.github.com

**Comment by nicovank** [comment]
I've made it my primary email, if it's not an option, `@fb`, `@gmail`, or the GitHub indirection is fine.

image

**Pull request merged**
